File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int n) {
        int[] ans = new int[n+1];
        Arrays.fill(ans, Integer.MAX_VALUE);
        ans[0] = 0;
        for (int deno : coins) {
          for (int amount = 0; amount < n + 1; amount++) {
            if (deno <= amount) {
              if (ans[amount - deno] != Integer.MAX_VALUE) {
              ans[amount] = Math.min(ans[amount], ans[amount - deno] + 1);
              }
            }
          }
        }
        return (ans[n] != Integer.MAX_VALUE) ? ans[n] : -1;
      }
}

File: 322-coin-change/322-coin-change.py
class Solution(object):
    def coinChange(self, coins, n):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        ans = [float("inf") for amount in range(n+1)]
        ans[0] = 0
        for deno in coins:
            for amount in range(len(ans)):
                if deno <= amount:
                    ans[amount] = min(ans[amount], ans[amount - deno] + 1)
        return ans[n] if ans[n] != float("inf") else -1